src/simd/generic.rs

```diff
@@ -68,35 +86,83 @@ pub trait SIMD<
-        {
-            #[cfg(target_arch = "x86")]
-            use std::arch::x86::_mm_prefetch;
-            #[cfg(target_arch = "x86_64")]
-            use std::arch::x86_64::_mm_prefetch;
-
-            _mm_prefetch(data as *const i8, 0); // 0=NTA
-        }
-        #[cfg(target_arch = "aarch64")]
-        {
-            use std::arch::aarch64::_prefetch;
-
-            _prefetch(data as *const i8, 0, 0); // 0=READ, 0=NTA
-        }
+    fn _get_overflow_lane_size_limit() -> usize {
+        Self::MAX_INDEX - Self::MAX_INDEX % LANE_SIZE
     }
+}
 
-    // ------------------------------------ ARGMINMAX --------------------------------------
+// ---------------------------------- SIMD algorithm -----------------------------------
 
-    unsafe fn argminmax(data: &[ScalarDType]) -> (usize, usize);
+// --------------- Default
 
+/// The default SIMDCore trait (for all data types)
+pub trait SIMDCore<ScalarDType, SIMDVecDtype, SIMDMaskDtype, const LANE_SIZE: usize>:
+    SIMDOps<ScalarDType, SIMDVecDtype, SIMDMaskDtype, LANE_SIZE>
+where
+    ScalarDType: Copy + PartialOrd + AsPrimitive<usize>,
+    SIMDVecDtype: Copy,
+    SIMDMaskDtype: Copy,
+{
     #[inline(always)]
-    unsafe fn _argminmax(data: &[ScalarDType]) -> (usize, usize)
-    where
-        SCALAR: ScalarArgMinMax<ScalarDType>,
-    {
-        argminmax_generic(data, LANE_SIZE, Self::_overflow_safe_core_argminmax)
+    unsafe fn _core_argminmax(arr: &[ScalarDType]) -> (usize, ScalarDType, usize, ScalarDType) {
+        assert_eq!(arr.len() % LANE_SIZE, 0);
+        // Efficient calculation of argmin and argmax together
+        let mut new_index = Self::INITIAL_INDEX;
+        let mut index_low = Self::INITIAL_INDEX;
+        let mut index_high = Self::INITIAL_INDEX;
+
+        let mut arr_ptr = arr.as_ptr(); // Array pointer we will increment in the loop
+        let mut values_low = Self::_mm_loadu(arr_ptr);
+        let mut values_high = Self::_mm_loadu(arr_ptr);
+
+        // This is (40%-5%) slower than the loop below (depending on the data type)
+        // arr.chunks_exact(LANE_SIZE)
+        //     .into_iter()
+        //     .skip(1)
+        //     .for_each(|step| {
+        //         new_index = Self::_mm_add(new_index, increment);
+
+        //         let new_values = Self::_mm_loadu(step.as_ptr());
+
+        //         let lt_mask = Self::_mm_cmplt(new_values, values_low);
+        //         let gt_mask = Self::_mm_cmpgt(new_values, values_high);
+
+        //         index_low = Self::_mm_blendv(index_low, new_index, lt_mask);
+        //         index_high = Self::_mm_blendv(index_high, new_index, gt_mask);
+
+        //         values_low = Self::_mm_blendv(values_low, new_values, lt_mask);
+        //         values_high = Self::_mm_blendv(values_high, new_values, gt_mask);
+        //     });
+
+        for _ in 0..arr.len() / LANE_SIZE - 1 {
+            // Increment the index
+            new_index = Self::_mm_add(new_index, Self::INDEX_INCREMENT);
```

Comment:
Performance question:

How does this look if we iterate on each of these arrays in separate for loops? Would that not increase our cache hits? We can't make assumptions about the locality of the data, but doing it this way means we're operating on what likely could be data with better locality.

This is for consideration, but if you do know the answer/have tried it, maybe throw in a comment explaining why that approach isn't faster than this.

Reply:
Very relevant question! 

To assure that we are on the same page - you are suggesting to instead of iterating over the entire array in a single loop, it might be more efficient to perform some sort of "chunked" iteration (e.g. 4*`LANE_SIZE`) in an inner loop? 
I can see how something like this can potentially decrease cache misses (as smaller "chunks"  that fully fit in cache can be reused in the inner loop).

Although I did not explore exactly what you described here, I did try loop unrolling - with no avail.

Reply:
Ah, let me clarify.

In this loop, we're reading from 3 different arrays at once for every step, with one for loop.

What I am suggesting *may* be faster is to iterate over each array separately (i.e. use 2x separate for-loops), as this would maximise our chances of getting cache & prefetch hits.

```rust
for _ in 0..arr.len() / LANE_SIZE - 1 {
    // do stuff on index high
    ...
}

for _ in 0..arr.len() / LANE_SIZE - 1 {
    // do stuff on index low
    ...
}
```

Reply:
Oh I see!

Guess assessing the potential performance gain of this can be done rather quickly - I'll analyze the cache misses some time next week :)